src/mds/MDCache.cc

```diff
@@ -5959,15 +5959,23 @@ bool MDCache::open_undef_inodes_dirfrags()
 	   << rejoin_undef_inodes.size() << " inodes "
 	   << rejoin_undef_dirfrags.size() << " dirfrags" << dendl;
 
-  set<CDir*> fetch_queue = rejoin_undef_dirfrags;
+  // dirfrag -> (fetch_complete, keys_to_fetch)
+  map<CDir*, pair<bool, std::vector<dentry_key_t> > > fetch_queue;
```

Comment:
Maybe the `fetch_queue` could just be:

`map<CDir*, std::vector<dentry_key_t>> fetch_queue;`

Just use `p.second.empty()` instead of `p.second.first` ?

Reply:
> Maybe the `fetch_queue` could just be:
> 
> `map<CDir*, std::vector<dentry_key_t>> fetch_queue;`
> 
> Just use `p.second.empty()` instead of `p.second.first` ?

The first here has meanings。True means fetching all keys in the cdir while false means fetch specified keys in the cdir.
p.second.empty() means only fetching the cdir head.